etcdserver/api/membership/cluster.go

```diff
@@ -252,6 +252,16 @@ func (c *RaftCluster) Recover(onSet func(*zap.Logger, *semver.Version)) {
 	}
 }
 
+// IsPromoteChange checks if m is a promoteChange
+func (c *RaftCluster) IsPromoteChange(m *Member) bool {
```

Comment:
`IsPromoteChange` needs not be exported?

And just return `return members[m.ID] != nil && members[m.ID].IsLearner && !m.IsLearner`?

Reply:
Also, can we document the logic here? (e.g. second member reconf call to the learner promotes...)

Reply:
+1

Reply:
+1

Reply:
> IsPromoteChange needs not be exported? 
the answer isno, if we want to use it in the (s *EtcdServer) applyConfChange,see
```
if s.cluster.IsPromoteChange(m) {
			s.cluster.UpdateRaftAttributes(m.ID, m.RaftAttributes)
		} else {
			s.cluster.AddMember(m)
		}

```

Reply:
> And just return return members[m.ID] != nil && members[m.ID].IsLearner && !m.IsLearner?

then we will need to get members from the outside the call.

Reply:
sure, more clear comment will be added